### ISSUE TITLE
Crash when trying to knap flint with a chalk stone

### ISSUE THREAD

**Issue by AiraPlaysGames** [anegostudios/VintageStory-Issues#1279]
**Game Version:** 1.15.5
**Platform:** Windows 10
**Modded:** No 

### Description
Game crashed while trying to do some knapping. Wasn't sure if I could use chalk to knap the flint but figured I'd try, and when I clicked the game crashed and gave me a crash log (attached below).

### Logs
```
Running on 64 bit Windows with 16 GB RAM 
Version: v1.15.5 (Stable)17-Aug-21 17:38:39: Critical error occurred
System.IndexOutOfRangeException: Index was outside the bounds of the array.
   at Vintagestory.GameContent.BlockEntityKnappingSurface.OnRemove(Vec3i voxelPos, Int32 radius)
   at Vintagestory.GameContent.BlockEntityKnappingSurface.OnUseOver(IPlayer byPlayer, Vec3i voxelPos, BlockFacing facing, Boolean mouseMode)
   at Vintagestory.GameContent.ItemStone.OnHeldAttackStop(Single secondsPassed, ItemSlot slot, EntityAgent byEntity, BlockSelection blockSel, EntitySelection entitySel)
   at Vintagestory.Client.NoObf.SystemMouseInWorldInteractions.HandleHandInteraction(Single dt)
   at Vintagestory.Common.EventManager.TriggerGameTick(Int64 ellapsedMilliseconds, IWorldAccessor world)
   at Vintagestory.Client.NoObf.ClientMain.MainRenderLoop(Single dt)
   at Vintagestory.Client.NoObf.ClientMain.MainGameLoop(Single deltaTime)
   at _EMxDKvc7iN7J6hlNFQiKeoxg0Mm._CEmopzLEk14BSqH9plcDEDfmuwE(Single )
   at _Wwq7msw90Af4XU7mxkxAUKyrS5B._iIGaTXNE8VYTpgEu9zjpAQyJSxd(Single )
   at _Wwq7msw90Af4XU7mxkxAUKyrS5B._vCB99EztL7cSfmplK8tyTwrQQxC(Single )
   at Vintagestory.Client.NoObf.ClientPlatformWindows.window_RenderFrame(Object sender, FrameEventArgs e)
   at System.EventHandler`1.Invoke(Object sender, TEventArgs e)
   at OpenTK.GameWindow.RaiseRenderFrame(Double elapsed, Double& timestamp)
   at OpenTK.GameWindow.DispatchRenderFrame()
   at OpenTK.GameWindow.Run(Double updates_per_second, Double frames_per_second)
   at _q065sSmA3CNgGuKTGnlAjR4ljPw._WkSl7yfiyr5DwPiSMvTudNAhp6(_t3ILnyYBq5OoYhhBY3JeTWgFmJDA , String[] )
   at _DBu5wiQWRAHGS5G2jiCKymvQ9yE._WkSl7yfiyr5DwPiSMvTudNAhp6(ThreadStart )
-------------------------------

Event Log entries containing Vintagestory.exe, the latest 3
==================================
{ TimeGenerated = 9-Aug-21 22:20:57, Site = , Source = Application Hang, Message = The program Vintagestory.exe version 1.15.3.0 stopped interacting with Windows and was closed. To see if more information about the problem is available, check the problem history in the Security and Maintenance control panel.

Process ID: a04

Start Time: 01d78d5b119da20a

Termination Time: 24

Application Path: C:\Users\Secret Foxfire\AppData\Roaming\Vintagestory\Vintagestory.exe

Report Id: a2dff6f8-1e84-4770-99a6-d2075b3f6b26

Faulting package full name: 

Faulting package-relative application ID: 

Hang type: Unknown
 }
--------------
{ TimeGenerated = 9-Aug-21 22:20:57, Site = , Source = Windows Error Reporting, Message = Fault bucket 1627544781492218016, type 5
Event Name: AppHangB1
Response: Not available
Cab Id: 0

Problem signature:
P1: Vintagestory.exe
P2: 1.15.3.0
P3: 61051c0c
P4: c2ee
P5: 134217728
P6: 
P7: 
P8: 
P9: 
P10: 

Attached files:
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WERAD1F.tmp.WERInternalMetadata.xml
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WERAD4F.tmp.xml
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WERAD4E.tmp.csv
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WERAD6E.tmp.txt
\\?\C:\Users\Secret Foxfire\AppData\Local\Temp\WER204.tmp.appcompat.txt

These files may be available here:
\\?\C:\ProgramData\Microsoft\Windows\WER\ReportArchive\AppHang_Vintagestory.exe_6a9d2cc7929738723063470f6c0d81d8b949cee_c6d645e5_6d6e02d4-a42a-4776-870e-3d8db3f6ba11

Analysis symbol: 
Rechecking for solution: 0
Report Id: a2dff6f8-1e84-4770-99a6-d2075b3f6b26
Report Status: 268435456
Hashed bucket: 38b758f2675487d84696335b0ce268a0
Cab Guid: 0 }
--------------
{ TimeGenerated = 24-Jul-21 19:46:30, Site = , Source = Windows Error Reporting, Message = Fault bucket 1601164709659358829, type 5
Event Name: RADAR_PRE_LEAK_64
Response: Not available
Cab Id: 0

Problem signature:
P1: Vintagestory.exe
P2: 1.14.10.0
P3: 10.0.19043.2.0.0
P4: 
P5: 
P6: 
P7: 
P8: 
P9: 
P10: 

Attached files:
\\?\C:\Users\SECRET~1\AppData\Local\Temp\RDR3CDD.tmp\empty.txt
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WER3CEE.tmp.WERInternalMetadata.xml
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WER3CFE.tmp.xml
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WER3D19.tmp.csv
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WER3D39.tmp.txt

These files may be available here:


Analysis symbol: 
Rechecking for solution: 0
Report Id: 88a36663-05a8-4085-b93e-c36edb1eb8f8
Report Status: 268435456
Hashed bucket: e2510997d6d277c626387ad1f223366d
Cab Guid: 0 }
```

**Comment by tyronx** [comment]
Cannot reproduce this :(

**Comment by copygirl** [comment]
It does still happen in 1.15.7, confirmed by another player: #1393

**Comment by tyronx** [comment]
Unfortunately I'm still not able to reproduce this.

**Comment by tyronx** [comment]
Does this happen in single player or multiplayer?

**Comment by AiraPlaysGames** [comment]
For me it happened on single player. It has not happened again since making the report but I also haven't tried to do it again.

**Comment by noelle-lavenza** [comment]
I can reproduce it, I think. The issue is that the radius is set with `slot.Itemstack.Collectible.GetToolMode` which is also set by cave art. That means it can sometimes have a radius large enough to cause issues.

**Comment by noelle-lavenza** [comment]
https://github.com/anegostudios/vssurvivalmod/blob/c1961caf8fecb4a463f34a345206058decde7205/BlockEntity/BEKnappingSurface.cs#L312
`                    if (voxelPos.X >= 0 && voxelPos.X < 16 && voxelPos.Z >= 0 && voxelPos.Z < 16 && Voxels[offPos.X, offPos.Z])`
This should probably be doing a bounds check on `offPos` rather than `voxelPos`.

**Comment by tyronx** [comment]
thanks for the research, this should be fixed in 1.18
